alf/environments/suite_babyai.py

```diff
@@ -0,0 +1,219 @@
+    VOCAB = [
+        'then',
+        'after',
+        'you',
+        'and',
+        'go',
+        'to',
+        'pick',
+        'up',
+        'open',
+        'put',
+        'next',
+        'door',
+        'ball',
+        'box',
+        'key',
+        'on',
+        'your',
+        'left',
+        'right',
+        'in',
+        'front',
+        'of',
+        'you',
+        'behind',
+        'red',
+        'green',
+        'blue',
+        'purple',
+        'yellow',
+        'grey',
+        'the',
+        'a',
+    ]
+
+    def __init__(self,
+                 env,
+                 max_instruction_length=32,
+                 one_token_per_step=False):
+        """
+        Args:
+            gym_env (gym.Env): An instance of OpenAI gym environment.
+            max_instruction_length (int): the maximum number of words of an instruction.
+            one_token_per_step (bool): If False, the whole instruction (word ID array)
+                is given in one step in the observation. If True, the word IDs are
+                given in the observation sequentially. Each step only one world ID
+                is given. Zeros are given for every steps after all the world IDs
+                are given.
+        """
+        super().__init__(env)
+
+        self._max_instruction_length = max_instruction_length
+        self._one_token_per_step = one_token_per_step
+
+        # the extra 1 is for padding
+        vocab_size = len(self.VOCAB) + 1
+
+        obs_space = {
+            'image':
+                env.observation_space['image'],
+            'direction':
+                gym.spaces.Discrete(4),
+            'mission':
+                gym.spaces.MultiDiscrete([vocab_size] * max_instruction_length)
+        }
+        if one_token_per_step:
+            obs_space['mission'] = gym.spaces.Discrete(vocab_size)
+
+        self.observation_space = gym.spaces.Dict(obs_space)
+
+        self._vocab = {'': 0}
+        for i, w in enumerate(self.VOCAB):
+            self._vocab[w] = i + 1
+        self._last_mission = ''
+        self._tokens = []
+        self._word_pattern = re.compile("([a-z]+)")
+
+    def _tokenize(self, instruction):
+        """Convert instruction string to a numpy array."""
+        tokens = self._word_pattern.findall(instruction.lower())
+        instr = np.array([self._vocab.get(token, 0) for token in tokens])
+        if np.amin(instr) == 0:
+            for token in tokens:
+                if token not in self._vocab:
+                    raise ValueError(
+                        "The instruction '%s' contains word "
+                        " out of vocabulary: %s" % (instruction, token))
+        if len(instr) < self._max_instruction_length:
+            instr = np.concatenate([
+                instr,
+                np.zeros([self._max_instruction_length - len(instr)],
+                         dtype=np.int64)
+            ])
+        elif len(instr) > self._max_instruction_length:
```

Comment:
How do you ensure the max instruction length is no greater than this number?

Reply:
I updated the default max length to 80. According to the grammar, the max length seems to be 72.